### PR TITLE
Provide a way for users to create efficient document-based fix all provider

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Alternate approach to https://github.com/dotnet/roslyn/pull/50223

This would let us get rid of https://github.com/dotnet/roslyn-analyzers/blob/master/src/Roslyn.Diagnostics.Analyzers/Core/FixAll/DocumentBasedFixAllProvider.cs

### CODE REVIEW

File: src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/FixAllProvider.cs [hunk 1]
```diff
@@ -37,5 +39,40 @@ public virtual IEnumerable<string> GetSupportedFixAllDiagnosticIds(CodeFixProvid
...
         /// </summary>
         public abstract Task<CodeAction?> GetFixAsync(FixAllContext fixAllContext);
+
+        /// <summary>
+        /// Create a <see cref="FixAllProvider"/> that fixes documents independently.  This should be used instead of
+        /// <see cref="WellKnownFixAllProviders.BatchFixer"/> in the case where fixes for a <see cref="Diagnostic"/>
+        /// only affect the <see cref="Document"/> the diagnostic was produced in.
+        /// </summary>
+        /// <param name="fixAllAsync">
+        /// Callback that will the diagnostics present in the provided document.  The document returned will only be
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/FixAllProvider.cs:49]:
```suggestion
        /// Callback that will the fix diagnostics present in the provided document.  The document returned will only be
```

File: src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/FixAllProvider.cs [hunk 2]
```diff
@@ -37,5 +39,40 @@ public virtual IEnumerable<string> GetSupportedFixAllDiagnosticIds(CodeFixProvid
...
+
+        /// <summary>
+        /// Create a <see cref="FixAllProvider"/> that fixes documents independently.  This should be used instead of
+        /// <see cref="WellKnownFixAllProviders.BatchFixer"/> in the case where fixes for a <see cref="Diagnostic"/>
+        /// only affect the <see cref="Document"/> the diagnostic was produced in.
+        /// </summary>
+        /// <param name="fixAllAsync">
+        /// Callback that will the fix diagnostics present in the provided document.  The document returned will only be
+        /// examined for its content (e.g. it's <see cref="SyntaxTree"/> or <see cref="SourceText"/>.  No other aspects
+        /// of (like it's properties), or changes to the <see cref="Project"/> or <see cref="Solution"/> it points at
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/FixAllProvider.cs:51]:
```suggestion
        /// of it (like attributes), or changes to the <see cref="Project"/> or <see cref="Solution"/> it points at
```

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/CodeFixes/SyntaxEditorBasedCodeFixProvider.FixAllProvider.cs [hunk 3]
```diff
@@ -1,52 +0,0 @@
-﻿// Licensed to the .NET Foundation under one or more agreements.
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/CodeFixes/SyntaxEditorBasedCodeFixProvider.FixAllProvider.cs:1]: removed now that we would have a public API we can use instead.

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/CodeFixes/SyntaxEditorBasedCodeFixProvider.FixAllProvider.cs [hunk 4]
```diff
@@ -1,52 +0,0 @@
...
-            private readonly SyntaxEditorBasedCodeFixProvider _codeFixProvider;
-
-            public SyntaxEditorBasedFixAllProvider(SyntaxEditorBasedCodeFixProvider codeFixProvider)
-                => _codeFixProvider = codeFixProvider;
-
-            protected override string GetFixAllTitle(FixAllContext fixAllContext)
-                => FixAllContextHelper.GetDefaultFixAllTitle(fixAllContext);
-
-            protected override async Task<Document?> FixAllAsync(FixAllContext fixAllContext, Document document, ImmutableArray<Diagnostic> diagnostics)
-            {
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/CodeFixes/SyntaxEditorBasedCodeFixProvider.FixAllProvider.cs:30]: moved into the containing type.

File: src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/FixAllProvider.cs [hunk 5]
```diff
@@ -37,5 +39,40 @@ public virtual IEnumerable<string> GetSupportedFixAllDiagnosticIds(CodeFixProvid
...
+        /// </param>
+        public static FixAllProvider Create(Func<FixAllContext, Document, ImmutableArray<Diagnostic>, Task<Document?>> fixAllAsync)
+        {
+            if (fixAllAsync == null)
+                throw new ArgumentNullException(nameof(fixAllAsync));
+
+            return new CallbackDocumentBasedFixAllProvider(fixAllAsync);
+        }
+
+        private class CallbackDocumentBasedFixAllProvider : AbstractDocumentBasedFixAllProvider
```
Comment (mavasani) [inline src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/FixAllProvider.cs:62]: Nit: 
```suggestion
        private sealed class CallbackDocumentBasedFixAllProvider : AbstractDocumentBasedFixAllProvider
```

**Review by mavasani — APPROVED** [review]

**Comment by CyrusNajmabadi** [comment]
Merged this into https://github.com/dotnet/roslyn/pull/50223.  Closing this out.

**Pull request closed without merging**